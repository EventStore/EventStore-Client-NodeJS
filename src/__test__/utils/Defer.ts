export class Defer {
  resolve!: () => void;
  reject!: (reason?: string | Error) => void;
  promise: Promise<void>;

  constructor() {
    this.promise = new Promise<void>((resolve, reject) => {
      this.reject = reject;
      this.resolve = resolve;
    });
  }
}
