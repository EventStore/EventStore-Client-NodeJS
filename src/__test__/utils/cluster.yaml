version: "3.5"

services:
  volumes-provisioner:
    image: hasnat/volumes-provisioner
    environment:
      PROVISION_DIRECTORIES: "1000:1000:0755:/tmp/certs"
    volumes:
      - "./certs:/tmp/certs"
    network_mode: none

  cert-gen:
    image: docker.pkg.github.com/eventstore/es-gencert-cli/es-gencert-cli:1.0.2
    entrypoint: bash
    command: >
      -c "es-gencert-cli create-ca -out /tmp/certs/ca &&
          es-gencert-cli create-node -ca-certificate /tmp/certs/ca/ca.crt -ca-key /tmp/certs/ca/ca.key -out \
          /tmp/certs/node1 -ip-addresses 127.0.0.1,172.30.240.11 &&
          es-gencert-cli create-node -ca-certificate /tmp/certs/ca/ca.crt -ca-key /tmp/certs/ca/ca.key -out \
          /tmp/certs/node2 -ip-addresses 127.0.0.1,172.30.240.12 &&
          es-gencert-cli create-node -ca-certificate /tmp/certs/ca/ca.crt -ca-key /tmp/certs/ca/ca.key -out \
          /tmp/certs/node3 -ip-addresses 127.0.0.1,172.30.240.13"
    user: "1000:1000"
    volumes:
      - "./certs:/tmp/certs"
    depends_on:
      - volumes-provisioner

  esdb-node1:
    image: docker.pkg.github.com/eventstore/eventstore/eventstore:ci
    env_file:
      - shared.env
    environment:
      - EVENTSTORE_GOSSIP_SEED=172.30.240.12:2113,172.30.240.13:2113
      - EVENTSTORE_INT_IP=172.30.240.11
      - EVENTSTORE_CERTIFICATE_FILE=/etc/eventstore/certs/node1/node.crt
      - EVENTSTORE_CERTIFICATE_PRIVATE_KEY_FILE=/etc/eventstore/certs/node1/node.key
      - EVENTSTORE_ADVERTISE_HOST_TO_CLIENT_AS=127.0.0.1
      - EVENTSTORE_ADVERTISE_HTTP_PORT_TO_CLIENT_AS=2111
      - EVENTSTORE_ADVERTISE_TCP_PORT_TO_CLIENT_AS=1111
      - EVENTSTORE_CLUSTER_SIZE=3
      - EVENTSTORE_RUN_PROJECTIONS=All
      - EVENTSTORE_INT_TCP_PORT=1112
      - EVENTSTORE_HTTP_PORT=2113
      - EVENTSTORE_TRUSTED_ROOT_CERTIFICATES_PATH=/etc/eventstore/certs/ca
      - EVENTSTORE_DISCOVER_VIA_DNS=false
      - EVENTSTORE_ENABLE_EXTERNAL_TCP=true
      - EVENTSTORE_ENABLE_ATOM_PUB_OVER_HTTP=true
    ports:
      - 1111:1113
      - 2111:2113
    networks:
      clusternetwork:
        ipv4_address: 172.30.240.11
    volumes:
      - ./certs:/etc/eventstore/certs
    restart: unless-stopped
    depends_on:
      - cert-gen

  esdb-node2:
    image: docker.pkg.github.com/eventstore/eventstore/eventstore:ci
    env_file:
      - shared.env
    environment:
      - EVENTSTORE_GOSSIP_SEED=172.30.240.11:2113,172.30.240.13:2113
      - EVENTSTORE_INT_IP=172.30.240.12
      - EVENTSTORE_CERTIFICATE_FILE=/etc/eventstore/certs/node2/node.crt
      - EVENTSTORE_CERTIFICATE_PRIVATE_KEY_FILE=/etc/eventstore/certs/node2/node.key
      - EVENTSTORE_ADVERTISE_HOST_TO_CLIENT_AS=127.0.0.1
      - EVENTSTORE_ADVERTISE_HTTP_PORT_TO_CLIENT_AS=2112
      - EVENTSTORE_ADVERTISE_TCP_PORT_TO_CLIENT_AS=1112
      - EVENTSTORE_CLUSTER_SIZE=3
      - EVENTSTORE_RUN_PROJECTIONS=All
      - EVENTSTORE_INT_TCP_PORT=1112
      - EVENTSTORE_HTTP_PORT=2113
      - EVENTSTORE_TRUSTED_ROOT_CERTIFICATES_PATH=/etc/eventstore/certs/ca
      - EVENTSTORE_DISCOVER_VIA_DNS=false
      - EVENTSTORE_ENABLE_EXTERNAL_TCP=true
      - EVENTSTORE_ENABLE_ATOM_PUB_OVER_HTTP=true
    ports:
      - 1112:1113
      - 2112:2113
    networks:
      clusternetwork:
        ipv4_address: 172.30.240.12
    volumes:
      - ./certs:/etc/eventstore/certs
    restart: unless-stopped
    depends_on:
      - cert-gen

  esdb-node3:
    image: docker.pkg.github.com/eventstore/eventstore/eventstore:ci
    env_file:
      - shared.env
    environment:
      - EVENTSTORE_GOSSIP_SEED=172.30.240.11:2113,172.30.240.12:2113
      - EVENTSTORE_INT_IP=172.30.240.13
      - EVENTSTORE_CERTIFICATE_FILE=/etc/eventstore/certs/node3/node.crt
      - EVENTSTORE_CERTIFICATE_PRIVATE_KEY_FILE=/etc/eventstore/certs/node3/node.key
      - EVENTSTORE_ADVERTISE_HOST_TO_CLIENT_AS=127.0.0.1
      - EVENTSTORE_ADVERTISE_HTTP_PORT_TO_CLIENT_AS=2113
      - EVENTSTORE_ADVERTISE_TCP_PORT_TO_CLIENT_AS=1113
      - EVENTSTORE_CLUSTER_SIZE=3
      - EVENTSTORE_RUN_PROJECTIONS=All
      - EVENTSTORE_INT_TCP_PORT=1112
      - EVENTSTORE_HTTP_PORT=2113
      - EVENTSTORE_TRUSTED_ROOT_CERTIFICATES_PATH=/etc/eventstore/certs/ca
      - EVENTSTORE_DISCOVER_VIA_DNS=false
      - EVENTSTORE_ENABLE_EXTERNAL_TCP=true
      - EVENTSTORE_ENABLE_ATOM_PUB_OVER_HTTP=true
    ports:
      - 1113:1113
      - 2113:2113
    networks:
      clusternetwork:
        ipv4_address: 172.30.240.13
    volumes:
      - ./certs:/etc/eventstore/certs
    restart: unless-stopped
    depends_on:
      - cert-gen

networks:
  clusternetwork:
    name: eventstoredb.local
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 172.30.240.0/24
