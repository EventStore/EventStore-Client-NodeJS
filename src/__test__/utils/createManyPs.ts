import {
  EventStoreDBClient,
  PersistentSubscriptionSettings,
  persistentSubscriptionToStreamSettingsFromDefaults,
  PersistentSubscriptionToAllSettings,
  persistentSubscriptionToAllSettingsFromDefaults,
} from "@eventstore/db-client";

export type CreatedPS = [
  streamName: string,
  groupName: string,
  settings: PersistentSubscriptionSettings
];

export const createManyPs =
  (client: EventStoreDBClient) =>
  async (
    streamName: (i: number) => string,
    groupName: (i: number) => string,
    settingsChanges: Partial<PersistentSubscriptionSettings>[]
  ): Promise<CreatedPS[]> => {
    const created: CreatedPS[] = [];

    for (const [i, settingsChange] of settingsChanges.entries()) {
      const toCreate: CreatedPS = [
        streamName(i),
        groupName(i),
        persistentSubscriptionToStreamSettingsFromDefaults(settingsChange),
      ];

      await client.createPersistentSubscriptionToStream(...toCreate);

      created.push(toCreate);
    }

    return created;
  };

export type CreatedPSToAll = [
  streamName: "$all",
  groupName: string,
  settings: PersistentSubscriptionToAllSettings
];

export const createManyPsToAll =
  (client: EventStoreDBClient) =>
  async (
    groupName: (i: number) => string,
    settingsChanges: Partial<PersistentSubscriptionToAllSettings>[]
  ): Promise<CreatedPSToAll[]> => {
    const created: CreatedPSToAll[] = [];

    for (const [i, settingsChange] of settingsChanges.entries()) {
      const toCreate: CreatedPSToAll = [
        "$all",
        groupName(i),
        persistentSubscriptionToAllSettingsFromDefaults(settingsChange),
      ];

      await client.createPersistentSubscriptionToAll(toCreate[1], toCreate[2]);

      created.push(toCreate);
    }

    return created;
  };
