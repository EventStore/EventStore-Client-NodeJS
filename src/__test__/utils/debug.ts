import createDebug from "debug";

export const testDebug = createDebug("esdb").extend("tests");
