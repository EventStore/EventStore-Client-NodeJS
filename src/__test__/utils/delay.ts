export const delay = (ms: number): Promise<void> =>
  new Promise((r) => setTimeout(r, ms));
