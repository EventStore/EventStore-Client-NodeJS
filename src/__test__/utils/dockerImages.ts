const esdbImage = ((): string => {
  const version = process.env.EVENTSTORE_IMAGE ?? "github:ci";
  switch (true) {
    case version.startsWith("local:"):
      return version.replace("local:", "");
    case version.startsWith("github:"):
      return version.replace(
        "github:",
        "ghcr.io/eventstore/eventstore/eventstore:"
      );
    case version.startsWith("dockerhub:"):
      return version.replace("dockerhub:", "eventstore/eventstore:");
    default:
      return version;
  }
})();

export const dockerImages = {
  volumesProvisioner: "hasnat/volumes-provisioner",
  certGen: "ghcr.io/eventstore/es-gencert-cli/es-gencert-cli:1.0.2",
  esdb: esdbImage,
};
