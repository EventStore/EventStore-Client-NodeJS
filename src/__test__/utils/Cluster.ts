import { join, relative } from "path";
import * as fs from "fs";
import { promisify } from "util";
import * as cp from "child_process";

import { v4 as uuid } from "uuid";
import * as getPort from "get-port";
import { upAll, down, exec, stopOne, logs } from "docker-compose/dist/v2";

import type { EndPoint, Certificate } from "../../types";

import { testDebug } from "./debug";
import { dockerImages } from "./dockerImages";

const rmdir = promisify(fs.rmdir);
const mkdir = promisify(fs.mkdir);
const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);
const CPExec = promisify(cp.exec);

type LogLevel = "Warning" | "Debug" | "Information" | "Error" | "Verbose";
interface LogLevels {
  Default: LogLevel;
  System: LogLevel;
  Microsoft: LogLevel;
  Grpc: LogLevel;
}

const nodeList = (count: number, ipStub: string) =>
  Promise.all<ClusterLocation>(
    Array.from({ length: count }, async (_, i) => ({
      port: await getPort(),
      ipv4_address: `${ipStub}.${10 + i}`,
    }))
  );

const createCertGen = (
  internalIPs: ClusterLocation[],
  domain: string,
  insecure: boolean
) =>
  insecure
    ? {}
    : {
        "cert-gen": {
          image: dockerImages.certGen,
          entrypoint: "bash",
          command: `-c "es-gencert-cli create-ca -out /tmp/certs/ca && ${internalIPs
            .map(
              ({ ipv4_address }, i) =>
                `es-gencert-cli create-node -ca-certificate /tmp/certs/ca/ca.crt -ca-key /tmp/certs/ca/ca.key -out /tmp/certs/node${i} -ip-addresses 127.0.0.1,${ipv4_address} -dns-names ${domain}`
            )
            .join(
              " && "
            )} && es-gencert-cli create-user -username admin -ca-certificate /tmp/certs/ca/ca.crt -ca-key /tmp/certs/ca/ca.key -out /tmp/certs/user-admin && es-gencert-cli create-user -username invalid -ca-certificate /tmp/certs/ca/ca.crt -ca-key /tmp/certs/ca/ca.key -out /tmp/certs/user-invalid && chown -R 1000:1000 /tmp/certs && chmod -R 755 /tmp/certs"`,
          user: "1000:1000",
          volumes: ["./certs:/tmp/certs"],
          depends_on: ["volumes-provisioner"],
        },
      };

interface ClusterLocation {
  port: number;
  ipv4_address: string;
}

const createNodes = (
  internalIPs: ClusterLocation[],
  domain: string,
  insecure: boolean,
  extraOptions: string[]
) =>
  internalIPs.reduce(
    (acc, { port, ipv4_address }, i, ipAddresses) => ({
      ...acc,
      [`esdb-node-${i}`]: {
        image: dockerImages.esdb,
        environment: [
          `EVENTSTORE_GOSSIP_SEED=${ipAddresses
            .reduce<string[]>(
              (acc, { ipv4_address: ip }) =>
                ip === ipv4_address ? acc : [...acc, `${ip}:2113`],
              []
            )
            .join(",")}`,
          `EVENTSTORE_INT_IP=${ipv4_address}`,
          `EVENTSTORE_ADVERTISE_HOST_TO_CLIENT_AS=${domain}`,
          `EVENTSTORE_ADVERTISE_HTTP_PORT_TO_CLIENT_AS=${port}`,
          `EVENTSTORE_CLUSTER_SIZE=${internalIPs.length}`,
          "EVENTSTORE_ENABLE_ATOM_PUB_OVER_HTTP=true",
          "EVENTSTORE_RUN_PROJECTIONS=All",
          "EVENTSTORE_LOG_CONFIG=/etc/eventstore/config/logconfig.json",
          "EVENTSTORE_DISCOVER_VIA_DNS=false",
          "EventStore__Plugins__UserCertificates__Enabled=true",
          ...(insecure
            ? [`EVENTSTORE_INSECURE=true`]
            : [
                `EVENTSTORE_CERTIFICATE_FILE=/etc/eventstore/certs/node${i}/node.crt`,
                `EVENTSTORE_CERTIFICATE_PRIVATE_KEY_FILE=/etc/eventstore/certs/node${i}/node.key`,
                "EVENTSTORE_TRUSTED_ROOT_CERTIFICATES_PATH=/etc/eventstore/certs/ca",
              ]),
          ...extraOptions,
        ],
        ports: [`${port}:2113`],
        networks: {
          clusternetwork: {
            ipv4_address,
          },
        },
        restart: "unless-stopped",
        ...(insecure
          ? {
              volumes: ["./config:/etc/eventstore/config"],
            }
          : {
              volumes: [
                "./certs:/etc/eventstore/certs",
                "./config:/etc/eventstore/config",
              ],
              depends_on: ["cert-gen"],
            }),
      },
    }),
    {}
  );

const rnd = (min: number, max: number) =>
  Math.floor(Math.random() * (max - min) + min);

export class Cluster {
  private id: string;
  private count: number;
  private insecure: boolean;
  private ipStub!: string;
  private certificates?: {
    root: Buffer;
    users: {
      admin: Certificate;
      invalid: Certificate;
    };
  };
  private retryCount = 3;
  private locations: ClusterLocation[] = [];
  private ready: Promise<void>;
  private inspected = false;
  private failed = false;
  private logLevel: LogLevels = {
    Default: "Debug",
    System: "Warning",
    Microsoft: "Warning",
    Grpc: "Warning",
  };
  private extraOptions: string[] = [];

  constructor(count: number, insecure = false, id = uuid()) {
    this.id = id;
    this.count = count;
    this.insecure = insecure;
    this.ready = this.initialize();
  }

  public setLogLevel = (levels: Partial<LogLevels>): this => {
    this.logLevel = { ...this.logLevel, ...levels };
    return this;
  };

  public setOption = (key: string, value: unknown): this => {
    this.extraOptions.push(`${key}=${value}`);
    return this;
  };

  public get certs() {
    if (this.insecure) {
      throw new Error("No Cert for insecure cluster");
    }
    if (this.failed || !this.certificates) {
      throw new Error("Cluster failed to initialize");
    }
    if (Object.values(this.certificates).some((value) => value === null)) {
      throw new Error("One or more certificate properties are null");
    }

    return this.certificates;
  }
  public get certPath() {
    if (this.insecure) throw new Error("No Cert for insecure cluster");

    return {
      root: this.path("./certs/ca/ca.crt"),
      admin: {
        certPath: this.path("./certs/user-admin/user-admin.crt"),
        certKeyPath: this.path("./certs/user-admin/user-admin.key"),
      },
      invalid: {
        certPath: this.path("./certs/user-invalid/user-invalid.crt"),
        certKeyPath: this.path("./certs/user-invalid/user-invalid.key"),
      },
    };
  }
  public domain = "client.bespin.dev";
  public get uri(): string {
    return `${this.domain}:${this.locations[0].port}`;
  }
  public get endpoints(): EndPoint[] {
    return this.locations.map(({ port }) => ({
      address: this.domain,
      port,
    }));
  }

  public up = async (): Promise<void> => {
    await this.ready;

    try {
      const { exitCode: e } = await upAll({ cwd: this.path() });
      if (e !== 0) throw `Exited with code ${e}`;
    } catch (error) {
      if (this.retryCount > 0) {
        this.retryCount -= 1;

        testDebug(
          `Failed to initialize cluster. Retry ${3 - this.retryCount}\n${
            error.err
          } `
        );

        await this.down();
        await this.cleanUp();
        await this.initialize();
        return this.up();
      }

      this.failed = true;

      console.error(`Cluster failed to initialize with error: ${error.err}`);
    }

    await this.healthy();

    if (this.count > 1) {
      await this.leaderElected();
    }

    if (!this.insecure) {
      this.certificates = {
        root: await readFile(this.certPath.root),
        users: {
          admin: {
            certFile: await readFile(this.certPath.admin.certPath),
            certKeyFile: await readFile(this.certPath.admin.certKeyPath),
          },
          invalid: {
            certFile: await readFile(this.certPath.invalid.certPath),
            certKeyFile: await readFile(this.certPath.invalid.certKeyPath),
          },
        },
      };
    }
  };

  public down = async (): Promise<void> => {
    if (this.inspected) {
      const composeFile = relative(
        process.cwd(),
        this.path("./docker-compose.yaml")
      );

      const folder = relative(process.cwd(), this.path());

      console.log(
        "Leaving cluster open for inspection. Dont forget to take it down: \n",
        `docker compose -f ${composeFile} down --volumes && rm -rf ${folder}`
      );

      return;
    }

    const { exitCode } = await down({
      cwd: this.path(),
      commandOptions: ["--volumes"],
    });

    if (exitCode !== 0) throw `Exited with code ${exitCode}`;

    await this.cleanUp();
  };

  public enableNodeLogs = async (
    node: EndPoint = this.endpoints[0]
  ): Promise<void> => {
    const nodeId = this.endpointToNodeId(node);
    logs(nodeId, {
      cwd: this.path(),
      follow: true,
      log: true,
    });
  };

  public killNode = async (node: EndPoint): Promise<void> => {
    const nodeId = this.endpointToNodeId(node);
    const response = await stopOne(nodeId, {
      cwd: this.path(),
    });

    if (response.exitCode === 0) return;

    console.log(response);
    throw new Error("Failed to kill node");
  };

  public resurrect = async (): Promise<void> => {
    const response = await upAll({
      cwd: this.path(),
    });

    if (response.exitCode === 0) {
      return this.healthy();
    }

    console.log(response);
    throw new Error("Failed to resurrect node");
  };

  public inspect = () => this.openInBrowser(false);
  public openInBrowser = async (launch = true): Promise<void> => {
    this.inspected = true;
    const url = `http${this.insecure ? "" : "s"}://${this.uri}`;

    if (launch) {
      console.log(`Opening ui: ${url}`);

      const start =
        process.platform == "darwin"
          ? "open"
          : process.platform == "win32"
          ? "start"
          : "xdg-open";

      await CPExec(start + " " + url);
    } else {
      console.log(`ui available at: ${url}`);
    }
  };

  private initialize = async (): Promise<void> => {
    this.ipStub = `172.${rnd(0, 255)}.${process.env.JEST_WORKER_ID ?? 1}`;
    this.locations = await nodeList(this.count, this.ipStub);

    const config = {
      version: "3.5",
      services: {
        "volumes-provisioner": {
          image: dockerImages.volumesProvisioner,
          environment: {
            PROVISION_DIRECTORIES: "1000:1000:0755:/tmp/certs",
          },
          volumes: ["./certs:/tmp/certs"],
          network_mode: "none",
        },
        ...createCertGen(this.locations, this.domain, this.insecure),
        ...createNodes(
          this.locations,
          this.domain,
          this.insecure,
          this.extraOptions
        ),
      },
      networks: {
        clusternetwork: {
          name: `${this.id}.eventstoredb.local`,
          driver: "bridge",
          ipam: {
            driver: "default",
            config: [
              {
                subnet: `${this.ipStub}.0/24`,
              },
            ],
          },
        },
      },
    };

    await mkdir(this.path(), { recursive: true });
    await mkdir(this.path("./config"), { recursive: true });
    await writeFile(
      this.path("./config/logconfig.json"),
      JSON.stringify({
        Logging: {
          LogLevel: this.logLevel,
        },
      })
    );
    await writeFile(this.path("./docker-compose.yaml"), JSON.stringify(config));
  };

  private healthy = async (...nodes: string[]) => {
    nodes = !nodes.length
      ? Array.from({ length: this.count }, (_, i) => `esdb-node-${i}`)
      : nodes;

    const healthy = new Set();

    while (healthy.size !== nodes.length) {
      for (const node of nodes) {
        if (healthy.has(node)) continue;

        try {
          const response = await exec(
            node,
            `curl --fail --insecure http${
              this.insecure ? "" : "s"
            }://localhost:2113/health/live`,
            { cwd: this.path() }
          );

          if (response.exitCode === 0) {
            healthy.add(node);

            testDebug(`--> ${node} is healthy`);
          }
        } catch (error) {
          // retry
        }
      }
    }
  };

  private leaderElected = async (...nodes: string[]) => {
    nodes = !nodes.length
      ? Array.from({ length: this.count }, (_, i) => `esdb-node-${i}`)
      : nodes;

    const ready = new Set();

    while (ready.size !== nodes.length) {
      for (const node of nodes) {
        if (ready.has(node)) continue;

        try {
          const response = await exec(
            node,
            `curl --fail --insecure http${
              this.insecure ? "" : "s"
            }://localhost:2113/gossip`,
            { cwd: this.path() }
          );

          interface Member {
            state: "Unknown" | "Leader" | "Follower";
          }

          const { members } = JSON.parse(response.out) as { members: Member[] };

          if (
            members.every(({ state }) => state !== "Unknown") &&
            members.some(({ state }) => state === "Leader")
          ) {
            ready.add(node);

            testDebug(`--> ${node} is ready`);
          }
        } catch (error) {
          // retry
        }
      }
    }
  };

  private path = (...parts: string[]): string => {
    return join(__dirname, "instances", this.id, ...parts);
  };

  private cleanUp = async () => {
    try {
      await rmdir(this.path(), { recursive: true });
    } catch (error) {
      if (process.env.CI) return;
      if (error.code === "ENOENT") return;

      console.log(
        `Failed to clean up test files at "${this.path()}": ${error}`
      );
    }
  };

  private endpointToNodeId = (endpoint: EndPoint) => {
    const index = this.locations.findIndex(
      (location) => location.port === endpoint.port
    );

    if (index === -1) {
      throw new Error(`unknown node ${endpoint.address}:${endpoint.port}`);
    }

    return `esdb-node-${index}`;
  };
}
