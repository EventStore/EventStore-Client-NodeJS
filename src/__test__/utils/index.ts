import { Cluster } from "./Cluster";

export const createTestNode = (): Cluster => new Cluster(1);
export const createTestCluster = (count = 3): Cluster => new Cluster(count);
export const createInsecureTestNode = (): Cluster => new Cluster(1, true);
export const createInsecureTestCluster = (count = 3): Cluster =>
  new Cluster(count, true);

export { Cluster };
export * from "./collect";
export * from "./Defer";
export * from "./delay";
export * from "./testEvents";
export * from "./optionalDescribe";
export * from "./postEventViaHttpApi";
