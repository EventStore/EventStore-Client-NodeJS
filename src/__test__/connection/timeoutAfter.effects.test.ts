import { EventStoreDBClient } from "../..";
import { createTestNode, jsonTestEvents } from "../utils";

describe("timeoutAfter", () => {
  const node = createTestNode();

  beforeAll(async () => {
    await node.up();
  });

  afterAll(async () => {
    await node.down();
  });

  describe("should time out a call", () => {
    test.each([
      [
        "client settings",
        () =>
          new EventStoreDBClient(
            { endpoint: node.uri, timeoutAfter: 1 },
            { rootCertificate: node.rootCertificate }
          ).appendToStream("client_settings", jsonTestEvents()),
      ],
      [
        "call options",
        () =>
          new EventStoreDBClient(
            { endpoint: node.uri },
            { rootCertificate: node.rootCertificate }
          ).appendToStream("call_options", jsonTestEvents(), {
            timeoutAfter: 1,
          }),
      ],
      [
        "call options override",
        () =>
          new EventStoreDBClient(
            { endpoint: node.uri, timeoutAfter: 20000 },
            { rootCertificate: node.rootCertificate }
          ).appendToStream("call_options_override", jsonTestEvents(), {
            timeoutAfter: 1,
          }),
      ],
    ])("%s", async (_, makeCall) => {
      await expect(makeCall).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
