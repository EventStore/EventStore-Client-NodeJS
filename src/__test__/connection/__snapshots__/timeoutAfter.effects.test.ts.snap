// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeoutAfter should time out a call call options 1`] = `"4 DEADLINE_EXCEEDED: Deadline exceeded"`;

exports[`timeoutAfter should time out a call call options override 1`] = `"4 DEADLINE_EXCEEDED: Deadline exceeded"`;

exports[`timeoutAfter should time out a call client settings 1`] = `"4 DEADLINE_EXCEEDED: Deadline exceeded"`;
